src/Lecture2.hs

```diff
@@ -215,7 +286,13 @@ The algorithm of merge sort is the following:
 [1,2,3]
 -}
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
+mergeSort [] = []
+mergeSort [x] = [x]
+mergeSort ls = merge sortLeft sortRight
+    where
+      (left, right) = splitAt (length ls `div` 2) ls
```

Comment:
You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.

Reply:
I see but let me clarify my understanding. In the current solution, the `length` function will traverse the list once to know the element counting, and `splitAt` will run through the list to the position desired. Thus, there would be almost two traversing on the same list.
If I split a list as follows, there is only one traverse.
```
(left, right) = simpleSplit 0 ls ([], [])
simpleSplit :: Int -> [Int] -> ([Int], [Int]) -> ([Int], [Int])
simpleSplit _ [] (evens, odds) = (evens, odds)
simpleSplit p (x:xs) (evens, odds)
    | odd p = simpleSplit (p+1) xs (evens, x : odds)
    | otherwise = simpleSplit (p+1) xs (x : evens, odds)
```

Reply:
Your understanding is correct 😌 
And you can write the function even shorter using pattern-matching capabilities in Haskel:

```haskell
(left, right) = simpleSplit ls

simpleSplit :: [Int] -> ([Int], [Int])
simpleSplit [] = ([], [])
simpleSplit [x] = ([x], [])
simpleSplit (x : y : rest) =
    let (xs, ys) = simpleSplit rest
    in (x : xs, y : ys)
```

Reply:
🤩 Elegance!